### PR TITLE
Changed UserInterface with identity + improved unit tests and doc

### PR DESCRIPTION
**Pull request by ezimuel**
This PR changes the `getUsername` of `UserInterface` in `getIdentity()`, as suggested in #14.

This PR contains also the following update:
- improvement on error management of `UserRepository\PdoDatabase::authenticate` and `UserRepository\PdoDatabase::getRolesFromUser`;
- improvement of the doc with the `UserRepository` adapter configuration values;
- improvement of the unit test with the Exception throw for the new error management of `UserRepository\PdoDatabase`.

**Note:** since this PR will change UserInterface it's a BC break. It must be released in 0.3. Moreover, this change will affect some `zend-expressive-authentication` adapters, like `zend-expressive-authentication-session`. We need to refactor these adapters as well.

### CODE REVIEW

File: test/UserRepository/PdoDatabaseTest.php [hunk 1]
```diff
@@ -20,32 +20,44 @@ public function testConstructor()
...
+        $pdoDatabase = new PdoDatabase($pdo, $this->getConfig());
 
         $user = $pdoDatabase->authenticate('test', 'password');
         $this->assertInstanceOf(UserInterface::class, $user);
-        $this->assertEquals('test', $user->getUsername());
+        $this->assertEquals('test', $user->getIdentity());
+    }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\RuntimeException
```
Comment (michalbundyra) [inline test/UserRepository/PdoDatabaseTest.php]: I would suggest to use method `expectException` just before line where the exception should be thrown, otherwise it could be thrown in any line of this test...

File: test/UserRepository/PdoDatabaseTest.php [hunk 2]
```diff
@@ -69,36 +75,63 @@ public function testAuthenticateInvalidUsername()
...
 
         $user = $pdoDatabase->authenticate('test', 'password');
         $this->assertInstanceOf(UserInterface::class, $user);
-        $this->assertEquals('test', $user->getUsername());
+        $this->assertEquals('test', $user->getIdentity());
         $this->assertEquals(['user', 'admin'], $user->getUserRoles());
     }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\RuntimeException
```
Comment (michalbundyra) [inline test/UserRepository/PdoDatabaseTest.php]: as above, please use method just before line where the exception should be thrown

File: test/UserRepository/PdoDatabaseTest.php [hunk 3]
```diff
@@ -69,36 +75,63 @@ public function testAuthenticateInvalidUsername()
...
+        $pdo = new PDO('sqlite:'. __DIR__ . '/../TestAssets/pdo_roles.sqlite');
+        $config = $this->getConfig();
+
+        $pdoDatabase = new PdoDatabase($pdo, $config);
+        $roles = $pdoDatabase->getRolesFromUser('foo');
+        $this->assertEmpty($roles);
+    }
+
+    /**
+     * @expectedException Zend\Expressive\Authentication\Exception\InvalidConfigException
```
Comment (michalbundyra) [inline test/UserRepository/PdoDatabaseTest.php]: the same

File: test/UserRepository/PdoDatabaseTest.php [hunk 4]
```diff
@@ -20,32 +20,44 @@ public function testConstructor()
...
     {
-        $pdo = new PDO('sqlite:'. __DIR__ . '/../TestAssets/pdo.sqlite');
-        $pdoDatabase = new PdoDatabase($pdo, [
+        return [
             'table' => 'user',
             'field' => [
-                'username' => 'username',
-                'password' => 'password',
+                'identity' => 'username',
+                'password' => 'password'
```
Comment (michalbundyra) [inline test/UserRepository/PdoDatabaseTest.php]: please re-add trailing comma

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,15 +2,16 @@
...
 
 ### Added
 
 - Nothing.
 
 ### Changed
 
-- Nothing
+- [#14](https://github.com/zendframework/zend-expressive-authentication/issues/14)
+  changed the `getUsername` with `getIdentity` in UserInterface.
```
Comment (weierophinney) [inline CHANGELOG.md]: Change to:

```markdown
renames the method `UserInterface::getUsername()` to `UserInterface::getIdentity()`.
```

Depending on another suggested change below, we may also need to detail something about the return value change.

File: docs/book/v1/intro.md [hunk 6]
```diff
@@ -22,11 +22,11 @@ namespace Zend\Expressive\Authentication;
 interface UserInterface
 {
     /**
-     * Get the username
+     * Get the unique user identity (id, username, email address or ...)
      *
      * @return string
      */
-    public function getUsername(): string;
+    public function getIdentity(): string;
```
Comment (weierophinney) [inline docs/book/v1/intro.md:29]: My understanding of #14 is that the reporter was also requesting that this method could return more than just a string - an object, for instance? If that's the case, we'd also need to remove the return type here, and modify the `@return` annotation to reflect that.

That said, I'd prefer we just have it return a string, and then the individual `UserInterface` implementations could provide other details via additional properties or methods.

File: src/UserRepository/PdoDatabase.php [hunk 7]
```diff
@@ -41,14 +42,19 @@ public function __construct(PDO $pdo, array $config)
...
             $this->config['table'],
-            $this->config['field']['username']
+            $this->config['field']['identity']
         );
 
         $stmt = $this->pdo->prepare($sql);
-        $stmt->bindParam(':username', $credential);
+        if (false === $stmt) {
+            throw new Exception\RuntimeException(
+                "Error during the user authentication, please check the configuration"
```
Comment (weierophinney) [inline src/UserRepository/PdoDatabase.php]: I'd change this to "An error occurred when preparing to fetch user details from the repository; please verify your configuration." This will give more information to the developer, without giving security details out should the exception message leak to a consumer.

**Review by weierophinney — CHANGES_REQUESTED** [review]

File: docs/book/v1/intro.md (hunk 6 above)
Comment (geerteltink) [reply docs/book/v1/intro.md:29]: > Get the unique user identity (id, username, email address or ...)

With _id_ I meant an integer or an UUID even. But I guess using `return (string) $this->id;` would solve the return type.

File: test/UserRepository/PdoDatabaseTest.php (hunk 1 above)
Comment (ezimuel) [reply test/UserRepository/PdoDatabaseTest.php]: Good catch, thanks!

File: docs/book/v1/intro.md (hunk 6 above)
Comment (ezimuel) [reply docs/book/v1/intro.md:29]: @weierophinney and @xtreamwayz ok, so we'll stay with a `string` return.

File: src/UserRepository/PdoDatabase.php (hunk 7 above)
Comment (ezimuel) [reply src/UserRepository/PdoDatabase.php]: @weierophinney good catch, thanks!

**Comment by ezimuel** [comment]
@weierophinney and @webimpress I updated the code with all your comments. Thanks for reviewing it.

File: src/UserRepository/HtpasswdFactory.php [hunk 8]
```diff
@@ -17,9 +17,7 @@ class HtpasswdFactory
      */
     public function __invoke(ContainerInterface $container) : Htpasswd
     {
-        $config = $container->has('config') ? $container->get('config') : [];
-        $htpasswd = $config['authentication']['htpasswd'] ?? null;
-
+        $htpasswd = $container->get('config')['authentication']['htpasswd'] ?? null;
```
Comment (weierophinney) [inline src/UserRepository/HtpasswdFactory.php:20]: Why this particular change?

File: src/UserRepository/PdoDatabase.php [hunk 9]
```diff
@@ -64,20 +71,32 @@ public function authenticate(string $credential, string $password = null) : ?Use
...
             );
         }
 
-        $stmt = $this->pdo->prepare($this->config['sql_get_roles']);
-        $stmt->bindParam(':username', $username);
+        try {
+            $stmt = $this->pdo->prepare($this->config['sql_get_roles']);
+        } catch (PDOException $e) {
+            throw new Exception\RuntimeException(sprintf(
+                "Error on sql_get_rols configuration: %s",
```
Comment (weierophinney) [inline src/UserRepository/PdoDatabase.php]: Two things on this one:

- Use single quotes (for consistency)
- I'd rephrase this to 'Error preparing retrieval of user roles: %s'; using "prepare" here helps the developer identify what sort of action was being attempted.

File: src/UserRepository/PdoDatabase.php [hunk 10]
```diff
@@ -64,20 +71,32 @@ public function authenticate(string $credential, string $password = null) : ?Use
...
+                $e->getMessage()
+            ));
+        }
+        if (false === $stmt) {
+            throw new Exception\RuntimeException(
+                "Error on sql_get_rols configuration"
+            );
```
Comment (weierophinney) [inline src/UserRepository/PdoDatabase.php]: Emit exactly the same exception message here as above, as this particular conditional is testing the same thing (error with statement preparation), but with PDO exceptions disabled. The only change is that the `%s` would become `unknown error`.

**Review by weierophinney — CHANGES_REQUESTED** [review]
Found a few more changes for you...

File: src/UserRepository/PdoDatabase.php (hunk 10 above)
Comment (michalbundyra) [reply src/UserRepository/PdoDatabase.php]: So maybe it should be moved into `Exception\RuntimeException` class with static method and parameter for `%s` as we already have in other expressive components?

File: docs/book/v1/intro.md (hunk 6 above)
Comment (michalbundyra) [reply docs/book/v1/intro.md:29]: We are using single space before colon in return type declaration (consistency)

File: docs/book/v1/user-repository.md [hunk 11]
```diff
@@ -42,3 +42,80 @@ result will be a `UserInterface` instance, otherwise a null value is returned.
...
+
+## htpasswd configuration
+
+The configuration of the `htpasswd` adapter requires only the file path, as
+follows:
+
+```php
+'config' => [
+    'authentication' => [
+        'htpasswd' => 'insert the path to htpasswd file'
```
Comment (michalbundyra) [inline docs/book/v1/user-repository.md]: I would add everywhere trailing commas in arrays in the docs as well (this is going to be a part of new coding-standard and it would be good to have it also in configuration, because folks could copy from there)

File: src/UserRepository/HtpasswdFactory.php (hunk 8 above)
Comment (ezimuel) [reply src/UserRepository/HtpasswdFactory.php:20]: Just because it's less line of code and we don't need `$config`.

**Comment by ezimuel** [comment]
@weierophinney and @webimpress incorporated the latest comments. Thanks again.

File: src/UserRepository/HtpasswdFactory.php (hunk 8 above)
Comment (weierophinney) [reply src/UserRepository/HtpasswdFactory.php:20]: The previous version wouldn't fail immediately if the `config` service was not present, but, ultimately, it would fail anyways, so I'm okay with the change.

**Review by weierophinney — APPROVED** [review]

File: src/UserRepository/HtpasswdFactory.php (hunk 8 above)
Comment (michalbundyra) [reply src/UserRepository/HtpasswdFactory.php:20]: @ezimuel @weierophinney wait... shouldn't container throw an exception (ContainerException) when service is not defined?
Please bare in mind we are mocking container always in tests, so not sure if it's right change.

Comment (michalbundyra) [reply src/UserRepository/HtpasswdFactory.php:20]: Ofc, check this out:

```php
    public function testContainer()
    {
        $container = new ServiceManager();
        $container->get('config');
    }
```

throws exception:
```
...Test::testContainer
Zend\ServiceManager\Exception\ServiceNotFoundException: Unable to resolve service "config" to a factory; are you certain you provided it during configuration?
```

Comment (weierophinney) [reply src/UserRepository/HtpasswdFactory.php:20]: Right, and that's what I was getting at. The change that @ezimuel makes here will now emit that particular exception if the `config` service is not defined. It raises another exception if the service is defined, but the expected configuration is missing. That seems like a reasonable change, as it actually gets more precise about the cause of failure.

Comment (michalbundyra) [reply src/UserRepository/HtpasswdFactory.php:20]: @weierophinney So we are missing that test case in tests now. By default container is prophesized to return null on all methods, but in that case exception should be thrown.

Comment (weierophinney) [reply src/UserRepository/HtpasswdFactory.php:20]: Okay, will add that test and push to @ezimuel 's branch shortly.
